File: src/shader.py
import glm
import numpy as np
from OpenGL.GL import *

from src.log import get_logger

logger = get_logger(__name__)
logger.setLevel(10)


class Shader:

    def __init__(self, vert_file: str, frag_file: str):
        self.id = glCreateProgram()
        logger.debug(f"Shader program id {self.id}")

        self.vertex_shader_id = Shader._compile_shader(vert_file, GL_VERTEX_SHADER)
        self.fragment_shader_id = Shader._compile_shader(frag_file, GL_FRAGMENT_SHADER)

        glAttachShader(self.id, self.vertex_shader_id)
        glAttachShader(self.id, self.fragment_shader_id)
        glLinkProgram(self.id)
        glValidateProgram(self.id)

        status = glGetProgramiv(self.id, GL_LINK_STATUS)

        if status == GL_FALSE:
            logger.critical("Shader program linking failed")
            print(glGetShaderInfoLog(self.id))
        else:
            logger.info("Shader program linked successfully")

    def use(self):
        glUseProgram(self.id)

    def stop(self):
        glUseProgram(0)

    def dispose(self):
        glDetachShader(self.id, self.vertex_shader_id)
        glDetachShader(self.id, self.fragment_shader_id)
        glDeleteShader(self.vertex_shader_id)
        glDeleteShader(self.fragment_shader_id)
        glDeleteProgram(self.id)

    def upload_uniform_float3(self, name: str, vector: glm.vec3):
        loc = self._get_uniform_location(name)
        glUniform3f(loc, vector.x, vector.y, vector.z)

    def upload_uniform_float2(self, name: str, vector: glm.vec2):
        loc = self._get_uniform_location(name)
        glUniform2f(loc, vector.x, vector.y)

    def upload_uniform_int1(self, name: str, value: int):
        loc = self._get_uniform_location(name)
        glUniform1i(loc, value)

    def upload_uniform_float16(self, name: str, matrix: glm.mat4):
        loc = self._get_uniform_location(name)
        glUniformMatrix4fv(loc, 1, GL_FALSE, np.array(matrix, dtype="float32"))  # TODO maybe it works?

    def _get_uniform_location(self, name: str) -> int:
        location = glGetUniformLocation(self.id, name)
        if location == -1:
            logger.error(f"Uniform variable '{name}' not found")

        return location

    @staticmethod
    def _compile_shader(source_file: str, type: int) -> int:
        with open(source_file, "r") as file:
            contents: str = file.read()

        id = glCreateShader(type)
        glShaderSource(id, contents)
        glCompileShader(id)

        status = glGetShaderiv(id, GL_COMPILE_STATUS)

        if status == GL_FALSE:
            logger.critical("Shader compilation failed")
            print(glGetShaderInfoLog(id))
        else:
            logger.info("Shader compiled successfully")

        return id

File: src/opengl/renderer/shader.cpp
#include <string>
#include <fstream>
#include <iostream>
#include <memory>

#include <glad/glad.h>
#include <glm/gtc/type_ptr.hpp>

#include "opengl/renderer/shader.h"
#include "other/logging.h"

Shader::Shader(GLuint program, GLuint vertex_shader, GLuint fragment_shader)
        : program(program), vertex_shader(vertex_shader), fragment_shader(fragment_shader) {
    SPDLOG_DEBUG("Created shader {}", program);
}

Shader::~Shader() {
    glDetachShader(program, vertex_shader);
    glDetachShader(program, fragment_shader);
    glDeleteShader(vertex_shader);
    glDeleteShader(fragment_shader);
    glDeleteProgram(program);

    SPDLOG_DEBUG("Deleted shader {}", program);
}

std::shared_ptr<Shader> Shader::create(const std::string& vertex_source_path,
                                       const std::string& fragment_source_path) {
    GLuint vertex_shader = compile_shader(vertex_source_path, GL_VERTEX_SHADER);
    GLuint fragment_shader = compile_shader(fragment_source_path, GL_FRAGMENT_SHADER);

    GLuint program = glCreateProgram();
    glAttachShader(program, vertex_shader);
    glAttachShader(program, fragment_shader);
    glLinkProgram(program);
    glValidateProgram(program);

    check_linking(program);

    return std::make_shared<Shader>(program, vertex_shader, fragment_shader);
}

void Shader::bind() const {
    glUseProgram(program);
}

void Shader::unbind() {
    glUseProgram(0);
}

void Shader::set_uniform_matrix(const std::string& name, const glm::mat4& matrix) const {
    GLint location = get_uniform_location(name);
    glUniformMatrix4fv(location, 1, GL_FALSE, glm::value_ptr(matrix));
}

void Shader::set_uniform_int(const std::string& name, int value) const {
    GLint location = get_uniform_location(name);
    glUniform1i(location, value);
}

GLint Shader::get_uniform_location(const std::string& name) const {
    GLint location = glGetUniformLocation(program, name.c_str());
    if (location == -1) {
        SPDLOG_ERROR("Uniform variable '{}' not found", name.c_str());
    }

    return location;
}    

GLuint Shader::compile_shader(const std::string& source_path, GLenum type) {
    std::ifstream file = std::ifstream(source_path.c_str(), std::ifstream::in);
    std::string source;

    if (file.is_open()) {
        std::string line;
        while (std::getline(file, line)) {
            source.append(line).append("\n");
        }
    } else {
        spdlog::critical("Could not open file '{}'", source_path.c_str());
        std::exit(1);
    }
    file.close();

    GLuint shader = glCreateShader(type);
    const char* s = source.c_str();
    glShaderSource(shader, 1, &s, nullptr);
    glCompileShader(shader);

    check_compilation(shader, type);

    return shader;
}

void Shader::check_compilation(GLuint shader, GLenum type) {
    GLint compile_status;
    glGetShaderiv(shader, GL_COMPILE_STATUS, &compile_status);

    if (compile_status == GL_FALSE) {
        GLsizei log_length;
        glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &log_length);

        const char* t;
        switch (type) {
            case GL_VERTEX_SHADER: t = "Vertex"; break;
            case GL_FRAGMENT_SHADER: t = "Fragment"; break;
        }

        if (log_length == 0) {
            spdlog::critical("{} shader compilation error with no message", t);
        } else {
            char* log_message = new char[log_length];
            glGetShaderInfoLog(shader, log_length, nullptr, log_message);

            spdlog::critical("{} shader compilation error\n{}", t, log_message);
            delete[] log_message;
        }

        std::exit(1);
    }
}

void Shader::check_linking(GLuint program) {
    GLint link_status;
    glGetProgramiv(program, GL_LINK_STATUS, &link_status);

    if (link_status == GL_FALSE) {
        GLsizei log_length;
        glGetProgramiv(program, GL_INFO_LOG_LENGTH, &log_length);

        if (log_length == 0) {
            spdlog::critical("Shader linking error with no message");
        } else {
            char* log_message = new char[log_length];
            glGetProgramInfoLog(program, log_length, nullptr, log_message);

            spdlog::critical("Shader linking error\n{}", log_message);
            delete[] log_message;
        }

        std::exit(1);
    }
}
